controller/api/destination/watcher/cluster_watcher.go

```diff
@@ -0,0 +1,320 @@
+}
+
+// newWatcherCacheWithDecoder is a helper function that allows the creation of a
+// cache with an arbitrary `configDecoder` function.
+func newClusterStoreWithDecoder(k8sAPI *k8s.API, enableEndpointSlices bool, decodeFn configDecoder) (*ClusterStore, error) {
+	cs := &ClusterStore{
+		store: make(map[string]cluster),
+		log: logging.WithFields(logging.Fields{
+			"component": "endpoints-watcher-cache",
+		}),
+		enableEndpointSlices: enableEndpointSlices,
+		k8sAPI:               k8sAPI,
+		decodeFn:             decodeFn,
+	}
+
+	_, err := cs.k8sAPI.Secret().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			secret, ok := obj.(*v1.Secret)
+			if !ok {
+				cs.log.Errorf("Error processing 'Secret' object: got %#v, expected *corev1.Secret", secret)
+				return
+			}
+
+			if secret.Type != consts.MirrorSecretType {
+				cs.log.Tracef("Skipping Add event for 'Secret' object %s/%s: invalid type %s", secret.Namespace, secret.Name, secret.Type)
+				return
+
+			}
+
+			clusterName, found := secret.GetLabels()[clusterNameLabel]
+			if !found {
+				cs.log.Tracef("Skipping Add event for 'Secret' object %s/%s: missing \"%s\" label", secret.Namespace, secret.Name, clusterNameLabel)
+				return
+			}
+
+			if err := cs.addWatcher(clusterName, secret); err != nil {
+				cs.log.Errorf("Error adding watcher for cluster %s: %v", clusterName, err)
+			}
+
+		},
+		DeleteFunc: func(obj interface{}) {
+			secret, ok := obj.(*v1.Secret)
+			if !ok {
+				// If the Secret was deleted when the watch was disconnected
+				// (for whatever reason) and the event was missed, the object is
+				// added with 'DeletedFinalStateUnknown'. Its state may be
+				// stale, so it should be cleaned-up.
+				tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+				if !ok {
+					cs.log.Debugf("unable to get object from DeletedFinalStateUnknown %#v", obj)
+					return
+				}
+				// If the zombie object is a `Secret` we can delete it.
+				secret, ok = tombstone.Obj.(*v1.Secret)
+				if !ok {
+					cs.log.Debugf("DeletedFinalStateUnknown contained object that is not a Secret %#v", obj)
+					return
+				}
+			}
+
+			clusterName, found := secret.GetLabels()[clusterNameLabel]
+			if !found {
+				cs.log.Tracef("Skipping Delete event for 'Secret' object %s/%s: missing \"%s\" label", secret.Namespace, secret.Name, clusterNameLabel)
+				return
+			}
+
+			cs.removeWatcher(clusterName)
+
+		},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return cs, nil
+}
+
+// Get safely retrieves a watcher from the cache given a cluster name. It
+// returns the watcher, cluster configuration, if the entry exists in the cache.
+func (cs *ClusterStore) Get(clusterName string) (*EndpointsWatcher, string, string, bool) {
+	cs.RLock()
+	defer cs.RUnlock()
+	s, found := cs.store[clusterName]
+	return s.watcher, s.trustDomain, s.clusterDomain, found
+}
+
+// GetWatcher is a convenience method that given a cluster name only returns the
```

Comment:
Do we actually need all of these convenience methods which return different subsets of the store entry?  In practice, we'll probably need all of it in the `destination.Get()` handler.  Maybe easier to expose fewer functions here (smaller API surface area) and let callers discard data if they don't need it.